keps/sig-node/20190717-seccomp-ga.md

```diff
@@ -0,0 +1,378 @@
+    // Load a profile defined in static file on the node.
+    // The profile must be preconfigured on the node to work.
+    // +optional
+    LocalhostProfile *string
+}
+
+type SeccompProfileType string
+
+const (
+    SeccompProfileUnconfined SeccompProfileType = "Unconfined"
+    SeccompProfileRuntime    SeccompProfileType = "Runtime"
+    SeccompProfileLocalhost  SeccompProfileType = "Localhost"
+)
+```
+
+This API makes the options more explicit than the stringly-typed annotation values, and leaves room
+for new profile sources to be added in the future (e.g. Kubernetes predefined profiles or ConfigMap
+profiles). The seccomp options struct leaves room for future extensions, such as defining the
+behavior when a profile cannot be set.
+
+#### PodSecurityPolicy API
+
+```go
+type PodSecurityPolicySpec struct {
+    ...
+    // seccomp is the strategy that will dictate allowable and default seccomp
+    // profiles for the pod.
+    // +optional
+    Seccomp *SeccompStrategyOptions
+    ...
+}
+
+type SeccompStrategyOptions struct {
+    // The default profile to set on the pod, if none is specified.
+    // The default MUST be allowed by the allowedProfiles.
+    // +optional
+    DefaultProfile *v1.SeccompProfile
+
+    // The set of profiles that may be set on the pod or containers.
+    // If unspecified, seccomp profiles are unrestricted by this policy.
+    // +optional
+    AllowedProfiles *SeccompProfileSet
+}
+
+// A set of seccomp profiles. This struct should be a plural of v1.SeccompProfile.
+// All values are optional, and an unspecified field excludes all profiles of
+// that type from the set.
+type SeccompProfileSet struct {
+    // The allowed seccomp profile types.
+    // +optional
+    Types []SeccompProfileType
+    // The allowed runtimeProfiles. A value of '*' allows all runtimeProfiles.
+    // +optional
+    RuntimeProfiles []string
+    // The allowed localhostProfiles. Values may end in '*' to include all
+    // localhostProfiles with a prefix.
+    // +optional
+    LocalhostProfiles []string
+}
+```
+
+## Design Details
+
+### Version Skew Strategy
+
+Because the API is currently represented as (mutable) annotations, care must be taken for migrating
+to the API fields. The cases to consider are: pod create, pod update, PSP create, PSP update.
+
+All API skew is resolved in the API server. New Kubelets will only use the seccomp values specified
+in the fields, and ignore the annotations.
+
+#### Pod Creation
+
+If no seccomp annotations or fields are specified, no action is necessary.
+
+If _only_ seccomp fields are specified, add the corresponding annotations. This ensures that the
+fields are enforced even if the node version trails the API version. Since [we
+support](https://kubernetes.io/docs/setup/release/version-skew-policy/) up to 2 minor releases of
+version skew between the master and node, this behavior will be removed in version N+2 (where N is
+the release with the upgraded seccomp support).
```

Comment:
update this to say "when the oldest supported kubelet is 1.18", since we're thinking through extending support to 4 releases

Reply:
Are we also considering extending version-skew support to 4 releases? Or just that we would carry patches for 4 releases?

Reply:
we're considering the latter. the current kubelet/apiserver skew was set so the oldest kubelet would work with the newest apiserver, to avoid needing to do multiple nodepool recreations in a single cluster upgrade cycle

Reply:
Actually, I think we should just keep this for at least 4 versions. We don't need to rush to remove the reconciliation.